dotnet/samples/Concepts/ChatCompletion/ChatHistoryReducers/ChatCompletionServiceExtensions.cs

```diff
@@ -0,0 +1,369 @@
+            .Select((value, index) => new { value, index })
+            .LastOrDefault(message => message.value.Metadata?.ContainsKey(SummaryMetadataKey) ?? false)
+            ?.index ?? -1;
+
+        var systemMessage = chatHistory.GetSystemMessage();
+        var hasSystemMessage = systemMessage is not null;
+
+        // check are there messages to be summarized
+        var startIndex = -1;
+        var endIndex = chatHistory.Count - this._truncatedSize;
+        if (lastIndex == -1)
+        {
+            // have never summarized so use chat history size
+            if (chatHistory.Count < this._summarizationThreshold)
+            {
+                return null;
+            }
+            startIndex = 0 + (hasSystemMessage ? 1 : 0);
+        }
+        else
+        {
+            // have summarized so use chat history size minus position of last summary
+            if (chatHistory.Count - lastIndex < this._summarizationThreshold)
+            {
+                return null;
+            }
+            startIndex = lastIndex;
+        }
+
+        var summaryMessage = await this.SummarizeAsync(chatHistory, startIndex, endIndex, cancellationToken);
+
+        // insert summary into the original chat history
+        chatHistory.Insert(endIndex + 1, summaryMessage);
+
+        IEnumerable<ChatMessageContent>? truncatedHistory = chatHistory.Extract(endIndex + 2, systemMessage: systemMessage, summaryMessage: summaryMessage);
+        return truncatedHistory;
+    }
+
+    #region private
+    /// <summary>
+    /// Summarize messages starting at the truncation index.
+    /// </summary>
+    /// <param name="chatHistory"></param>
+    /// <param name="startIndex"></param>
+    /// <param name="endIndex"></param>
+    /// <param name="cancellationToken"></param>
+    /// <returns></returns>
+    private async Task<ChatMessageContent> SummarizeAsync(ChatHistory chatHistory, int startIndex, int endIndex, CancellationToken cancellationToken)
+    {
+        // extract history for summarization
+        IEnumerable<ChatMessageContent> messagesToSummarize =
+            chatHistory.Extract(startIndex, endIndex: endIndex,
+                messageFilter: (m) => m.Items.Any(i => i is FunctionCallContent || i is FunctionResultContent));
+
+        // summarize the chat history
+        var summarizationRequest = new ChatHistory(this._summarizationPrompt);
+        summarizationRequest.AddRange(messagesToSummarize);
+        ChatMessageContent summaryContent = await this._chatClient.GetChatMessageContentAsync(summarizationRequest, cancellationToken: cancellationToken).ConfigureAwait(false);
+        summaryContent.Metadata = new Dictionary<string, object?> { { SummaryMetadataKey, true } };
+
+        return summaryContent;
+    }
+    #endregion
+}
+
+/// <summary>
+/// Interface for reducing the chat history before sending it to the chat completion provider.
+/// </summary>
+public interface IChatHistoryReducer
```

Comment:
Suggest moving to another file.

```suggestion
public interface IChatHistoryReducer
```